R/parse.R

```diff
@@ -52,17 +52,18 @@ parse_delim <- function(x, delim, quote = "\"", ...) {
 }
 
 parse_name_equals_value <- function(x) {
-  loc <- regexpr("=", x, fixed = TRUE)
-  pieces <- regmatches(x, loc, invert = TRUE)
+  if (length(x) == 0) return(NULL)
+  pieces <- strsplit(x, "=")
```

Comment:
What happens if there are multiple equals in the string? It doesn't look like that case is currently covered in the tests, so you might want to add a tests specifically for `parse_name_equals_value()` too.

Reply:
The only case it doesn't work is this type of scenario: `foo==&bar=test`. To overcome with we can do a few options:

Original Method:
```r
loc <- regexpr("=", x, fixed = TRUE)
regmatches(x, loc, invert = TRUE)
```

stringi method:
```r
stringi::stri_split_fixed(x, "=", n = 2)
```

substitution method:
```r
#i.e.
safe = "&"
strsplit(sub("=", safe, x, fixed = T), safe, fixed = TRUE)
```
**NOTE:** As `parse_name_equals_value` is called after `parse_delim` we can use the splitting delimiter in `parse_delim` as a "safe" character for the substitution method.


### Benchmark example:
``` r
safe <- "&"
query <- c(
  "foo==", "foo==&bar==", "foo==&bar==&cho==", "foo==&bar==&cho==&zoo=="
)
query_param <- lapply(query,
  httr2:::parse_delim, delim = safe
)

bms <- lapply(seq_along(query_param), \(i) {
  x <- query_param[[i]]
  bm <- bench::mark(
    original = {
      loc <- regexpr("=", x, fixed = TRUE)
      regmatches(x, loc, invert = TRUE)
    },
    stringi = stringi::stri_split_fixed(x, "=", n = 2),
    substitution = strsplit(sub("=", safe, x, fixed = T), safe, fixed = TRUE)
  )
  bm$param_no <- i
  bm
})

bms |> dplyr::bind_rows() |> ggplot2::autoplot()
#> Loading required namespace: tidyr
```

![](https://i.imgur.com/gJwyoK8.png)<!-- -->

<sup>Created on 2024-01-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

The substitution method has some really good performance with it only being slightly slower than `stringi`.